File: AXAPI/HaoConnect/php/HaoConnect/HaoConnect.php
<?php
/**
 * HaoConnect，
 * @package HaoConnect
 * @author axing
 * @since 1.0
 * @version 1.0
 */
require_once(__dir__.'/HaoHttpClient.php');
require_once(__dir__.'/HaoConfig.php');
require_once(__dir__.'/HaoUtility.php');
require_once(__dir__.'/HaoResult.php');

define('METHOD_GET','get');
define('METHOD_POST','post');

class HaoConnect {

    Public static $Clientinfo         = null;                           //应用信息
    Public static $Clientversion      = null;                           //使用本类所在客户端的版本号
    Public static $apiHost            = null;                           //接口域名
    Public static $SECRET_HAX_CONNECT = null;                           //加密秘钥，这里用的是2号设备类型的密钥


    Public static $Isdebug       = '0';                                  //是否打印调试信息

    Public static $Devicetype    = '2';                                  //设备类型 1：浏览器设备 2：pc设备 3：Android设备 4：ios设备 5：windows phone设备
    Public static $Devicetoken   = '0';                                  //推送用的设备token

    Public static $Requesttime   = '0';                                  //请求时的时间戳，单位：秒

    Public static $Userid        = '';                                  //当前用户ID，登录后可获得。
    Public static $Logintime     = '';                                  //登录时间，时间戳，单位：秒，数据来自服务器
    Public static $Checkcode     = '';                                  //Userid和Logintime组合加密后的产物，用于进行用户信息加密。数据来自服务器

    /**
     *  请求加密后的校验串，服务器会使用同样规则加密请求后，比较校验串是否一致，从而防止请求内容被纂改。
     *  取头信息里Clientversion,Devicetype,Requesttime,Devicetoken,Userid,Logintime,Checkcode,Clientinfo,Isdebug  和 表单数据
     *  每个都使用key=value（空则空字符串）格式组合成字符串然后放入同一个数组
     *  再放入请求地址（去除http://和末尾/?#之后的字符）后
     *  并放入私钥字符串后自然排序
     *  连接为字符串后进行MD5加密，获得Signature
     *  将Signature也放入头信息，进行传输。
     */
    Public static $Signature     = '0';//接口加密校验

    /** 更新用户信息（一般是从cookie中取） */
    public static function setCurrentUserInfo($userID='',$loginTime='',$checkCode='')
    {
        static::$Userid    = $userID;
        static::$Logintime = $loginTime;
        static::$Checkcode = $checkCode;
        //todo 存到cookie里
        HaoUtility::storeCookie('Userid',$userID);
        HaoUtility::storeCookie('Logintime',$loginTime);
        HaoUtility::storeCookie('Checkcode',$checkCode);
    }

    /** 更新用户的设备信息 */
    public static function setCurrentDeviceToken($deviceToken='')
    {
        static::$Devicetoken = $deviceToken;
        HaoUtility::storeCookie('Devicetoken',$deviceToken);
    }

    /** 根据链接、参数、头信息，取得加密后的密钥 */
    protected static function getSignature($actionUrl='',$params = array(),$headers = array())
    {
        //定义一个空的数组
        $tmpArr = array();

        if ($params == null)
        {
            $params = array();
        }

        //将所有头信息和数据组合成字符串格式：%s=%s，存入上面的数组
        foreach (array('Clientversion','Devicetype','Devicetoken','Requesttime','Userid','Logintime','Checkcode') as $_key) {
            if (array_key_exists($_key,$headers))
            {
                array_push($tmpArr, sprintf('%s=%s', $_key, $headers[$_key]));
            }
            else
            {
                throw new Exception('缺少头信息：'.$_key);
            }
        }

        if (abs($headers['Requesttime'] - time()) > 5*60 )//300
        {
            throw new Exception('请求的时间不正确，请检查网络或重试。');
        }

        //加密版本2.0，支持应用识别码和debug模式
        if (!isset($params['r']))
        {
            foreach (array('Clientinfo','Isdebug') as $_key) {
                if (array_key_exists($_key,$headers))
                {
                    array_push($tmpArr, sprintf('%s=%s', $_key, $headers[$_key]));
                }
                else
                {
                    throw new Exception('缺少加密方案2.0头信息：'.$_key);
                }
            }

            array_push($tmpArr, sprintf('%s=%s', 'link', preg_replace ('/^http.*?:\/\/(.*?)(\/*[\?#].*$|[\?#].*$|\/*$)/', '$1', $actionUrl)));
        }

        //同样的，将所有表单数据也组成字符串后，放入数组。（注：file类型不包含）
        foreach ($params as $_key => $_value) {
            if (!is_array($_value) && !is_object($_value))
            {
                array_push($tmpArr, sprintf('%s=%s', $_key, $_value));
            }
        }

        //最后，将一串约定好的密钥字符串也放入数组。（不同的项目甚至不同的版本中，可以使用不同的密钥）
        array_push($tmpArr, static::$SECRET_HAX_CONNECT);

        //对数组进行自然排序
        sort($tmpArr, SORT_STRING);

        //将排序后的数组组合成字符串
        $tmpStr = implode( $tmpArr );

        //对这个字符串进行MD5加密，即可获得Signature
        return md5( $tmpStr );
    }

    protected static function getSecretHeaders($urlParam,$params = array())
    {
        if (static::$Userid == '' && isset($_COOKIE['Userid']))
        {
            static::$Userid    = $_COOKIE['Userid'];
            static::$Logintime = $_COOKIE['Logintime'];
            static::$Checkcode = $_COOKIE['Checkcode'];
        }
        if (static::$Devicetoken == '' && isset($_COOKIE['Devicetoken']))
        {
            static::$Devicetoken = $_COOKIE['Devicetoken'];
        }

        if (static::$Clientinfo == null)
        {
            static::$Clientinfo              = HaoConfig::$Clientinfo;
            static::$Clientversion           = HaoConfig::$Clientversion;
            static::$SECRET_HAX_CONNECT      = HaoConfig::$SECRET_HAX_CONNECT;
            static::$apiHost                 = HaoConfig::$apiHost;
        }


        $headers = array();
        $headers['Clientinfo']    = static::$Clientinfo ;
        $headers['Clientversion'] = static::$Clientversion ;
        $headers['apiHost']       = static::$apiHost ;
        $headers['Isdebug']       = static::$Isdebug ;
        $headers['Devicetype']    = static::$Devicetype ;
        $headers['Devicetoken']   = static::$Devicetoken ;
        $headers['Requesttime']   = time();
        $headers['Userid']        = static::$Userid ;
        $headers['Logintime']     = static::$Logintime ;
        $headers['Checkcode']     = static::$Checkcode ;

        $actionUrl = sprintf('http://%s/%s',static::$apiHost , $urlParam);

        $headers['Signature']       = static::getSignature($actionUrl,$params,$headers);

        /** 作为转发服务器，也要转发用户的IP，声明自己的代理行为 */
        if ($headers['Clientversion']==2 && isset($_SERVER))
        {
            $headers['X_FORWARDED_FOR'] = $_SERVER['REMOTE_ADDR'];
            $headers['CLIENT_IP']       = $_SERVER['REMOTE_ADDR'];
        }

        return $headers;
    }

    /** 请求API地址，获得字符串 */
    public static function loadContent($urlParam, $params = array(), $method = METHOD_GET)
    {
    	/** 更新下用户信息（如果用户信息尚未更新的话） */

    	$headers = static::getSecretHeaders($urlParam,$params);

        $actionUrl = sprintf('http://%s/%s',static::$apiHost , $urlParam);

    	return HaoHttpClient::loadContent($actionUrl,$params,$method,$headers);
    }

    public static function loadJson($urlParam,  $params = array(),$method = METHOD_GET)
    {
        $content = static::loadContent($urlParam,$params,$method);
        try {
            return json_decode($content,true);
        } catch (Exception $e) {
            return $e;
        }
        return $content;
    }

    /**
     * 发起请求并返回HaoResult
     * @param  string $urlParam 接口方法
     * @param  array  $params   参数字典
     * @param  string $method   请求方式
     * @return HaoResult        返回结果
     */
    public static function request($urlParam,  $params = array(),$method = METHOD_GET)
    {
    	$content = static::loadContent($urlParam,$params,$method);
    	try {
            $tmpResult = json_decode($content,true);
            if ( isset($tmpResult['modelType']) && $tmpResult['modelType'] == 'HaoResult' )
            {
        		return HaoResult::instanceModel($tmpResult['results'],$tmpResult['errorCode'],$tmpResult['errorStr'],$tmpResult['extraInfo'],$tmpResult['resultCount']);
            }
    	} catch (Exception $e) {
            return HaoResult::instanceModel($content,-1,'数据解析失败，请联系管理员。',null);
        }
        return HaoResult::instanceModel($content,-1,'未解析到正确的数据，请联系管理员。',null);
    }

    /**
     * 发起GET请求并返回HaoResult
     * @param  string $urlParam 接口方法
     * @param  array  $params   参数字典
     * @return HaoResult        返回结果
     */
    public static function get($urlParam,  $params = array())
    {
        return static::request($urlParam,  $params, METHOD_GET);
    }


    /**
     * 发起POST请求并返回HaoResult
     * @param  string $urlParam 接口方法
     * @param  array  $params   参数字典
     * @return HaoResult        返回结果
     */
    public static function post($urlParam,  $params = array())
    {
        return static::request($urlParam,  $params, METHOD_POST);
    }

}

File: AXAPI/HaoConnect/java/HaoConnect/HaoConnect.java
import android.content.Context;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.loopj.android.http.AsyncHttpResponseHandler;
import com.loopj.android.http.RequestParams;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import cz.msebera.android.httpclient.Header;

/**
 * Created by wangtao on 15/12/21.
 */
public class HaoConnect {

    private static String Clientinfo = "";
    private static String Clientversion = "0.1";
    private static String SECRET_HAX_CONNECT = "";
    private static String ApiHost = "";

    private static String Devicetype = "3";
    private static String Devicetoken = "";
    private static String Requesttime = "0";

    private static String Userid = "";
    private static String Logintime = "";
    private static String Checkcode = "";

    public static String METHOD_GET = "get";
    public static String METHOD_POST = "post";

    /**
     * 是否开启调试模式
     */
    private static String Isdebug = "0";

    /**
     * 请求加密后的校验串，服务器会使用同样规则加密请求后，比较校验串是否一致，从而防止请求内容被纂改。
     * 取头信息里Clientversion,Devicetype,Requesttime,Devicetoken,Userid,Logintime,Checkcode,Clientinfo,Isdebug  和 表单数据
     * 每个都使用key=value（空则空字符串）格式组合成字符串然后放入同一个数组
     * 再放入请求地址（去除http://和末尾/?#之后的字符）后
     * 并放入私钥字符串后自然排序
     * 连接为字符串后进行MD5加密，获得Signature
     * 将Signature也放入头信息，进行传输。
     */
    private static String Signature = "";

    /**
     * 使用前必须调用该初始化方法
     */
    public static void init() {
        if (Clientinfo == null || Clientinfo.length() == 0) {
            Clientinfo = HaoConfig.HAOCONNECT_CLIENTINFO;
            SECRET_HAX_CONNECT = HaoConfig.HAOCONNECT_SECRET_HAX;
            ApiHost = HaoConfig.HAOCONNECT_APIHOST;
        }
    }

    /**
     * 设置版本号
     */
    public void setClientVersion(String clientVersion) {
        Clientversion = clientVersion;
    }

    /**
     * 设置用户相关信息
     * @param userID        用户id
     * @param loginTime     登录时间
     * @param checkCode
     */
    public static void setCurrentUserInfo(String userID, String loginTime, String checkCode) {
        Userid = userID;
        Logintime = loginTime;
        Checkcode = checkCode;

        HaoConfig.putString("userID", Userid);
        HaoConfig.putString("loginTime", Logintime);
        HaoConfig.putString("checkCode", Checkcode);
    }

    /**
     * 推送token
     * @param deviceToken
     */
    public static void setCurrentDeviceToken(String deviceToken) {
        Devicetoken = deviceToken;
    }

    /**
     * @param requestData
     * @param urlParam
     * @return 请求头数据，里面包括加密字段
     */
    private static Map<String, Object> getSecretHeaders(Map<String, Object> requestData, String urlParam) {
        Map<String, Object> headers = new HashMap<>();
        headers.put("Clientinfo", Clientinfo);
        headers.put("Clientversion", Clientversion);
        headers.put("Devicetype", Devicetype);
        headers.put("Requesttime", (System.currentTimeMillis() / 1000) + "");
        headers.put("Devicetoken", Devicetoken);
        headers.put("Isdebug", "0");

        if (Userid == null || Userid.equals(""))
        {
            Userid = HaoConfig.getString("userID");
            Logintime = HaoConfig.getString("loginTime");
            Checkcode = HaoConfig.getString("checkCode");
        }
        headers.put("Userid", Userid);
        headers.put("Logintime", Logintime);
        headers.put("Checkcode", Checkcode);

        Map<String, Object> signMap = new HashMap<>();
        signMap.putAll(headers);
        signMap.putAll(requestData);
        Map<String, Object> linkMap = new HashMap<String, Object>();
        linkMap.put("link", HaoUtility.httpStringFilter("http://" + ApiHost + "/" + urlParam));
        signMap.putAll(linkMap);
        headers.put("Signature", getSignature(signMap));
        return headers;
    }

    /**
     * 加密算法
     */
    private static String getSignature(Map<String, Object> map) {
        List<String> tmpArr = new ArrayList<String>();
        String secret = "";
        for (Map.Entry<String, Object> entry : map.entrySet()) {
            String data = entry.getKey() + "=" + entry.getValue();
            tmpArr.add(data);
        }
        tmpArr.add(SECRET_HAX_CONNECT);
        Collections.sort(tmpArr);
        for (String string : tmpArr) {
            secret += string;
        }
        return HaoUtility.encodeMD5String(secret);
    }

    public static void loadContent(String urlParam, Map<String, Object> params, String method, final HaoRequestResopnse haoRequestResopnse, Context context) {
        RequestParams requestParams = new RequestParams();
        for (Map.Entry<String, Object> entry : params.entrySet()) {
            requestParams.put(entry.getKey(), entry.getValue() + "");
        }
        HaoHttpClient.loadContent("http://" + ApiHost + "/" + urlParam, requestParams, method, getSecretHeaders(params, urlParam), new AsyncHttpResponseHandler() {
            @Override
            public void onSuccess(int i, Header[] headers, byte[] bytes) {
                String response = new String(bytes);
                haoRequestResopnse.requestOnSuccess(response);
            }

            @Override
            public void onFailure(int i, Header[] headers, byte[] bytes, Throwable throwable) {
                String response = new String(bytes);
                haoRequestResopnse.requestOnSuccess(response);
            }

            @Override
            public void onStart() {
                super.onStart();
                haoRequestResopnse.requestOnStart();
            }
        }, context);
    }

    public static void loadJson(String urlParam, Map<String, Object> params, String method, final HaoRequestResopnse haoRequestResopnse, Context context) {
        loadContent(urlParam, params, method, new HaoRequestResopnse() {
            @Override
            public void requestOnSuccess(Object result) {
                try {
                    Gson gson = new Gson();
                    JsonObject jsonObject = gson.fromJson(result.toString(), JsonObject.class);
                    haoRequestResopnse.requestOnSuccess(jsonObject);
                } catch (Exception e) {
                    haoRequestResopnse.requestOnFail("json解析失败，请检查返回结果");
                }
            }

            @Override
            public void requestOnStart() {
                haoRequestResopnse.requestOnStart();
            }

            @Override
            public void requestOnFail(Object results) {
                haoRequestResopnse.requestOnFail(results);
            }
        }, context);
    }

    public static void request(String urlParam, Map<String, Object> params, String method, final HaoConnectResponse haoConnectResponse, Context context) {

        loadJson(urlParam, params, method, new HaoRequestResopnse() {
            @Override
            public void requestOnSuccess(Object result) {
                try {
                    JsonObject jsonObject = (JsonObject) result;
                    HaoResult haoResult = (HaoResult) HaoResult.instanceModel(jsonObject.get("results"), jsonObject.get("errorCode").getAsInt(), jsonObject.get("errorStr").getAsString(), jsonObject.get("extraInfo"));

                    if (haoResult.isResultsOK()) {
                        haoConnectResponse.requestOnSuccess(haoResult);
                    } else {
                        haoConnectResponse.requestOnFail(haoResult);
                    }
                } catch (Exception e) {
                    HaoResult haoResult = (HaoResult) HaoResult.instanceModel(null, -1, e.toString(), null);
                    haoConnectResponse.requestOnFail(haoResult);
                }
            }

            @Override
            public void requestOnStart() {
                haoConnectResponse.requestOnStart();
            }

            @Override
            public void requestOnFail(Object results) {

                HaoResult haoResult = (HaoResult) HaoResult.instanceModel(null, -1, results.toString(), null);
                haoConnectResponse.requestOnFail(haoResult);
            }
        }, context);
    }

    public static void cancelRequest(Context context) {
        HaoHttpClient.cancelRequest(context);
    }

    public interface HaoRequestResopnse
    {
        void requestOnSuccess(Object result);

        void requestOnStart();

        void requestOnFail(Object results);
    }
}
